### DISCUSSION TITLE
Problems with training recognition model for vocab "German"

### DISCUSSION THREAD

**Discussion by tilman67** [mindee/doctr#1366] [Q&A] [answered]
Hi,

I tried to train an own version of the crnn_vgg16_bn recognition model with the vocab "german". I followed the instructions found in the docs and the code. In particular, I used the script references/recognition/train_pytorch.py (I use PyTorch) with the following parameters:

`--font <see_below>`
`--epochs 300`
`--vocab german`
`crnn_vgg16_bn`

The fonts were:

`LiberationSans-Regular.ttf`
`LiberationSans-Italic.ttf`
`LiberationSans-Bold.ttf`
`LiberationSans-BoldItalic.ttf`
`LiberationMono-Regular.ttf`
`LiberationMono-Italic.ttf`
`LiberationMono-Bold.ttf`
`LiberationMono-BoldItalic.ttf`
`LiberationSerif-Regular.ttf`
`LiberationSerif-Italic.ttf`
`LiberationSerif-Bold.ttf`
`LiberationSerif-BoldItalic.ttf`

(thus, I used "synthetic" training data created from fonts).

I then tested the new model with an ocr_predictor created as follows:

```
reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False, vocab=VOCABS["german"])
reco_params = load(<path_to_my_model_file>, map_location="cpu")
reco_model.load_state_dict(reco_params)
predictor = ocr_predictor(det_arch="db_resnet50", reco_arch=reco_model, pretrained=True)
```
However, the results were very poor. Many characters were not recognized correctly by the new model, whereas the original crnn_vgg16_bn model recognizes them almost perfectly.

What may be the reason for this? Did I made any mistake?

Any suggestions are greatly appreciated!

Best regards,
Tilman

**Comment by felixT2K** [comment]
Hi @tilman67 👋🏼,

The current `WordGenerator` needs some "improvements", it works well for debugging but to train on synthetic data i suggest to take a look at: https://github.com/clovaai/synthtiger

**Reply by tilman67** [reply]
Hi @felixT2K ,

thank you very much for the quick reply!

I'll follow your advice and give synthiger a try.

Best regards,
Tilman

**Reply by tilman67** [reply]
Hi @felixT2K ,

I trained with 80000 images generated with synthtiger (+ 20000 for validation). As input for synthtiger I used 100000 german words, a mix of real words from dictionaries, created numbers, and random sequences of vocab characters.

The results are better now. DocTR can recognize german umlauts quite well. However, there are still problems with "normal" characters. DocTR confuses characters, or cannot distinguish the upper and lower forms of characters.

Can you give me any advice how to fix the problems?

Is it necessary that the words used for training are real words? Or can I use arbitrary character sequences as well?

Is it important that the character frequencies in the words used for training reflect the real frequencies in the language? In my training, german umlauts (äöüÄÖÜß) were far more frequent then in the usual german language. I did that because the umlauts are the new thing docTR should learn, so a presented a lot of umlauts to it during training.

Best regards,
Tilman

**Reply by felixT2K** [reply, marked as answer]
Hi @tilman67 👋🏼,

Have you trained the model from scratch or used the `--pretrained` flag ? :)

You can also make use of arbitrary character sequences but the approach with real words is definitely not wrong.
The character frequency in a language doesn't matter because the model learns only on visual features so it has no "language understanding".

Another option would be to freeze the whole model and train only the last layer (classifier):

```
# Freeze all layers except the last one -> CRNN models
for name, param in model.named_parameters():
    if name in ["linear.weight", "linear.bias"]:
        param.requires_grad = True
        print(f"Set layer {name} to trainable")
        continue
    param.requires_grad = False
```
can be placed after the model init in the train script 

Best regards,
Felix :)

**Reply by tilman67** [reply]
Hi @felixT2K ,

it turned out  the `--pretrained` flag did the trick!

Thank you very much for your help!

Best regards,
Tilman :)

**Reply by felixT2K** [reply]
@tilman67 Nice :)

Are you interested to share your model with the community ? 🤗

**Reply by tilman67** [reply]
I'd love to to do so, but I have to ask the management of the company I'm working for for permission. I hope they agree!

**Reply by tilman67** [reply]
Hi @felixT2K ,

finally they agreed! To share the model with the community, I should follow [this instructions](https://mindee.github.io/doctr/using_doctr/sharing_models.html), right?

Best regards,
Tilman

**Reply by felixT2K** [reply]
@tilman67 Great to hear :)

Correct 👍🏼

**Reply by felixdittrich92** [reply]
Feel free to share the link to your model.
After some validation we can add it to the table in the docs :hugs:

**Reply by tilman67** [reply]
Hi @felixdittrich92 ,

I uploaded the model to Hugging Face: https://huggingface.co/tilman-rassy/doctr-crnn-vgg16-bn-fascan-v1.

Perhaps I should mention that I used a vocab that is not included in the VOCABS dict:
```
vocab = VOCABS["german"] + "àâéèêëîïôùûçÀÂÉÈÊËÎÏÔÙÛÇ" + "§"
```
(We wanted both german and french special characters to be recognized, as well as the § symbol.)

Best regards,
Tilman

**Reply by whoami02** [reply]
Hey @tilman67 @felixT2K , I am trying something similar, just want to include Ø symbol in recognition, when I added vocab the same way you have mentioned I got "ValueError: some characters cannot be found in 'vocab'"

This was the modification made by me:

```
vocab = VOCABS["english"]+ "Ø"
```

My labels include characters like 
```
    "001_10.jpg": "Ø5.2",
    "001_12.jpg": "4",
    "001_21.jpg": "11.25°",
    "001_22.jpg": "22.5°",
    "001_30.jpg": "Ø14.5",
```
Is it possbile that I am getting this error due to "Ø" symbol?
Can you help me?

**Reply by felixdittrich92** [reply]
Hi @whoami02 👋,

Could you share your train command and the modifications in datasets/vocabs.py please :)

Also check please that every char in your labels exists in the given vocab string

**Reply by whoami02** [reply]
Hi @felixdittrich92,
Thanks for the reply :)

here's my train command.
```
 python references/recognition/train_pytorch.py sar_resnet31 --train_path "E:\AW\OCR\fine-tune\Recog_data\train_set" --val_path "E:\AW\OCR\fine-tune\Recog_data\val_set" -b 16 --train-samples 127 --val-samples 25 --vocab "mathematical" --epochs 5 --wb --pretrained
```

and here's the most recent modification that I made
```
VOCABS['mathematical'] = "".join(
    dict.fromkeys(
        VOCABS["latin"]+
        VOCABS["english"]+
        "Ø±x"        
    )
)
```
My labels are basically digits, some alphabets, plus special characters "Ø and ±". My hunch is that these symbols might be wrongly interpreted, not sure though.
I have attached my labels.json file here
[labels.json](https://github.com/mindee/doctr/files/14370979/labels.json)

This is what I am getting, I have also followed this thread so far #1029 
```
(.venv) PS E:\AW\OCR\fine-tune\doctr> python references/recognition/train_pytorch.py sar_resnet31 --train_path "E:\AW\OCR\fine-tune\Recog_data\train_set" --val_path "E:\AW\OCR\fine-tune\Recog_data\val_set" -b 16 --train-samples 127 --val-samples 25 --vocab "english" --epochs 5 --wb --pretrained
Namespace(arch='sar_resnet31', train_path='E:\\AW\\OCR\\fine-tune\\Recog_data\\train_set', val_path='E:\\AW\\OCR\\fine-tune\\Recog_data\\val_set', train_samples=127, val_samples=25, font='FreeMono.ttf,FreeSans.ttf,FreeSerif.ttf', min_chars=1, max_chars=12, name=None, epochs=5, batch_size=16, device=None, input_size=32, lr=0.001, weight_decay=0, workers=None, resume=None, vocab='english', test_only=False, freeze_backbone=False, show_samples=False, wb=True, push_to_hub=False, pretrained=True, sched='cosine', amp=False, find_lr=False, early_stop=False, early_stop_epochs=5, early_stop_delta=0.01)
Validation set loaded in 0.0156s (25 samples in 2 batches)
WARNING:root:Invalid model URL, using default initialization.
Train set loaded in 0.0s (127 samples in 7 batches)
wandb: Currently logged in as: whoami02. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.16.3
wandb: Run data is saved locally in E:\AW\OCR\fine-tune\doctr\wandb\run-20240222_140054-wwskvfat
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run sar_resnet31_20240222-140052
wandb:  View project at https://wandb.ai/whoami02/text-recognition
wandb:  View run at https://wandb.ai/whoami02/text-recognition/runs/wwskvfat
  0%|                                                                                                                                | 0/7 [00:52<?, ?it/s]
Traceback (most recent call last):                                                                                                   | 0/7 [00:00<?, ?it/s]
  File "E:\.venv\Lib\site-packages\doctr\datasets\utils.py", line 76, in encode_string
    return list(map(vocab.index, input_string))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: substring not found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 488, in <module>
    main(args)
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 393, in main
    fit_one_epoch(model, train_loader, batch_transforms, optimizer, scheduler, amp=args.amp)
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 122, in fit_one_epoch
    train_loss = model(images, targets)["loss"]
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\torch\nn\modules\module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\torch\nn\modules\module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\doctr\models\recognition\sar\pytorch.py", line 245, in forward
    _gt, _seq_len = self.build_target(target)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\doctr\models\recognition\core.py", line 35, in build_target
    encoded = encode_sequences(sequences=gts, vocab=self.vocab, target_size=self.max_length, eos=len(self.vocab))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\doctr\datasets\utils.py", line 151, in encode_sequences
    for idx, seq in enumerate(map(partial(encode_string, vocab=vocab), sequences)):
  File "E:\.venv\Lib\site-packages\doctr\datasets\utils.py", line 78, in encode_string
    raise ValueError("some characters cannot be found in 'vocab'")
ValueError: some characters cannot be found in 'vocab'
....
```

**Reply by whoami02** [reply]
So I saw this in one of the discussions and tried it to check encodings

```
from doctr.datasets.utils import encode_sequences
from doctr.datasets import VOCABS
x = encode_sequences(["Ø5.2", "15/EN-80", "30°"], vocab=VOCABS['mathematical'])
print(x)
```
Output: 
```
[[100   5  75   2  -1  -1  -1  -1  -1]
 [  1   5  76  40  49  74   8   0  -1]
 [  3   0  94  -1  -1  -1  -1  -1  -1]]
```

I also tried to synthesize text image for, 'Ø,0,O,o' (dia, zero, upper and lower case O) as these seem very much alike
```
import matplotlib.pyplot as plt
from doctr.datasets.generator.base import synthesize_text_img
plt.imshow(synthesize_text_img("Ø0Oo")); plt.show()
```
Output :
![Figure_1](https://github.com/mindee/doctr/assets/43184560/03e8af54-1dd2-4d3e-8175-46fa9b4804b0)

**Reply by felixdittrich92** [reply]
@whoami02 

In your labels are whitespaces 

```
file: 001_22.jpg, character:   not in vocabulary, label:  22.5°
file: 001_67.jpg, character:   not in vocabulary, label:  3.Stress
file: 001_85.jpg, character:   not in vocabulary, label:  500037,
file: 001_136.jpg, character:   not in vocabulary, label:  0091-40-23079523
file: 001_139.jpg, character:   not in vocabulary, label:  to
file: 001_152.jpg, character:   not in vocabulary, label:  0.015B
file: 001_154.jpg, character:   not in vocabulary, label:  Ø31
file: 001_161.jpg, character:   not in vocabulary, label:  Ø205
file: 001_162.jpg, character:   not in vocabulary, label:  Ø201
file: 001_175.jpg, character:   not in vocabulary, label:  15°
file: 001_180.jpg, character:   not in vocabulary, label:  130/Ø129.9
file: 001_186.jpg, character:   not in vocabulary, label:  03-11-2020
file: 001_193.jpg, character:   not in vocabulary, label:  05-11-2020
file: 001_200.jpg, character:   not in vocabulary, label:  0.015B
file: 001_202.jpg, character:   not in vocabulary, label:  0.015B
file: 001_210.jpg, character:   not in vocabulary, label:  Ø160
file: 001_214.jpg, character:   not in vocabulary, label:  5.02/Ø
```

We don't support whitespace recognition and whitespace is also not in the vocab so if you remove all the whitespaces from your labels it should work ;)

Corrected:

[labels.json](https://github.com/mindee/doctr/files/14373833/labels.json)

**Reply by whoami02** [reply]
Hey @felixdittrich92 

I manually double-checked for whitespaces, also used the labels file you've attached.
Still getting the same error.

**Reply by felixdittrich92** [reply]
Could you set a `print` in File "E:\.venv\Lib\site-packages\doctr\datasets\utils.py", line 78, in encode_string and post the output please ? 
```
def encode_string(
    input_string: str,
    vocab: str,
) -> List[int]:
    """Given a predefined mapping, encode the string to a sequence of numbers

    Args:
    ----
        input_string: string to encode
        vocab: vocabulary (string), the encoding is given by the indexing of the character sequence

    Returns:
    -------
        A list encoding the input_string
    """
    try:
        return list(map(vocab.index, input_string))
    except ValueError as e:
        print(e)
        print("vocab:", vocab)
        print("input_string:", input_string)
        raise ValueError("some characters cannot be found in 'vocab'")
```


I have tested it shortly with your labels.json file without any issues after cleaning the whitespaces :thinking:

**Reply by felixdittrich92** [reply]
@whoami02 I see the issue in your train command you need to pass `--vocab=mathematical` instead of `--vocab=english`

**Reply by whoami02** [reply]
Like you said, I was using english vocab in the training command.
When I changed it to --vocab=mathematical, I am now getting KeyError: mathematical 
```
(.venv) PS E:\AW\OCR\fine-tune\doctr> python references/recognition/train_pytorch.py vitstr_base --train_path "E:\AW\OCR\fine-tune\Recog_data\train_set" --val_path "E:\AW\OCR\fine-tune\Recog_data\val_set" -b 16 --train-samples 127 --val-samples 25 --vocab=mathematical --epochs 5 --wb --pretrained
Namespace(arch='vitstr_base', train_path='E:\\AW\\OCR\\fine-tune\\Recog_data\\train_set', val_path='E:\\AW\\OCR\\fine-tune\\Recog_data\\val_set', train_samples=127, val_samples=25, font='FreeMono.ttf,FreeSans.ttf,FreeSerif.ttf', min_chars=1, max_chars=12, name=None, epochs=5, batch_size=16, device=None, input_size=32, lr=0.001, weight_decay=0, workers=None, resume=None, vocab='mathematical', test_only=False, freeze_backbone=False, show_samples=False, wb=True, push_to_hub=False, pretrained=True, sched='cosine', amp=False, find_lr=False, early_stop=False, early_stop_epochs=5, early_stop_delta=0.01)
Traceback (most recent call last):
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 488, in <module>
    main(args)
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 189, in main
    vocab = VOCABS[args.vocab]
            ~~~~~~^^^^^^^^^^^^
KeyError: 'mathematical'
```

Looking at this I updated my mathematical vocab inside vocab.py of the venv site packages. It solved the KeyError:
Also printed the vocab perfectly. But still the same error due to substring.

```
(.venv) PS E:\AW\OCR\fine-tune\doctr> python references/recognition/train_pytorch.py vitstr_base --train_path "E:\AW\OCR\fine-tune\Recog_data\train_set" --val_path "E:\AW\OCR\fine-tune\Recog_data\val_set" -b 16 --train-samples 127 --val-samples 25 --vocab=mathematical --epochs 5 --wb --pretrained
Namespace(arch='vitstr_base', train_path='E:\\AW\\OCR\\fine-tune\\Recog_data\\train_set', val_path='E:\\AW\\OCR\\fine-tune\\Recog_data\\val_set', train_samples=127, val_samples=25, font='FreeMono.ttf,FreeSans.ttf,FreeSerif.ttf', min_chars=1, max_chars=12, name=None, epochs=5, batch_size=16, device=None, input_size=32, lr=0.001, weight_decay=0, workers=None, resume=None, vocab='mathematical', test_only=False, freeze_backbone=False, show_samples=False, wb=True, push_to_hub=False, pretrained=True, sched='cosine', amp=False, find_lr=False, early_stop=False, early_stop_epochs=5, early_stop_delta=0.01)
Validation set loaded in 1.691s (25 samples in 2 batches)
WARNING:root:Invalid model URL, using default initialization.
Train set loaded in 0.0157s (127 samples in 7 batches)
wandb: Currently logged in as: whoami02. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.16.3
wandb: Run data is saved locally in E:\AW\OCR\fine-tune\doctr\wandb\run-20240222_184455-n1s9kggp
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run vitstr_base_20240222-184453
wandb:  View project at https://wandb.ai/whoami02/text-recognition
wandb:  View run at https://wandb.ai/whoami02/text-recognition/runs/n1s9kggp
                                                                                                                           substring not found                                                                                   | 0/7 [00:00<?, ?it/s]
vocab: 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~°£€¥¢฿Ø±
input_string: Ã˜118
  0%|                                                                                                | 0/7 [00:40<?, ?it/s]
Traceback (most recent call last):
  File "E:\.venv\Lib\site-packages\doctr\datasets\utils.py", line 76, in encode_string
    return list(map(vocab.index, input_string))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: substring not found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 488, in <module>
    main(args)
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 393, in main
    fit_one_epoch(model, train_loader, batch_transforms, optimizer, scheduler, amp=args.amp)
  File "E:\AW\OCR\fine-tune\doctr\references\recognition\train_pytorch.py", line 122, in fit_one_epoch
    train_loss = model(images, targets)["loss"]
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\torch\nn\modules\module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\torch\nn\modules\module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\doctr\models\recognition\vitstr\pytorch.py", line 86, in forward
    _gt, _seq_len = self.build_target(target)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\doctr\models\recognition\vitstr\base.py", line 31, in build_target
    encoded = encode_sequences(
              ^^^^^^^^^^^^^^^^^
  File "E:\.venv\Lib\site-packages\doctr\datasets\utils.py", line 154, in encode_sequences
    for idx, seq in enumerate(map(partial(encode_string, vocab=vocab), sequences)):
  File "E:\.venv\Lib\site-packages\doctr\datasets\utils.py", line 81, in encode_string
    raise ValueError("some characters cannot be found in 'vocab'")
ValueError: some characters cannot be found in 'vocab'
```

**Reply by felixdittrich92** [reply]
OK this sounds to me that you haven't installed doctr in editable mode

```
cd doctr
pip3 install -e .[torch] # or [tf] or [dev]
```

**Reply by felixdittrich92** [reply]
Afterwards save the files again and try :) 
Now it should also find the new created vocab without trouble

**Reply by whoami02** [reply]
Okay, I will try that in a fresh virtual env and let you know :)

**Reply by felixdittrich92** [reply]
Btw i saw that you want to use sar_resnet31 (there is still a bug with inference) i suggest to use parseq instead it's smaller, faster and more accurate :)

**Reply by whoami02** [reply]
I have tried what you told me to, installed everything freshly in editable mode.
I am still getting the above error. After printing the vocab like you mentioned. I found that the input string is taking something like this:
```
vocab: 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~°£€¥¢฿Ø±×
input_string: Ã˜45.3
```
As you can see its not in the vocab that's why I am probably getting the error. 
In order to be sure, I took all the keys in labels.json and added them into VOCAB["mathematical"], and yet the vocab remained same.
Hence the labels do not have anything outside of the given vocab.
I do not understand from where these characters come into the input string.

PS - I even added multilingual vocab along with my mathematical vocab to see if the error still persists, it still returned the same error

here's another run
```
vocab: 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~°£€¥¢฿Ø±×
input_string: 130/Ã˜129.9
```

Hey is it possible that different models are interpreting images differently? 
The previous commands were run with vitstr_base, while the last one was with crnn_mobilenet_v3_large

**Reply by whoami02** [reply]
Hey @felixdittrich92!
It finally started. 

After replying you I realized that the system was reading Ø as Ã˜ and ° as Â°, which is weird as I had used standard encodings while adding them. 
So when I added Â° and Ã˜ into the vocab, it worked.
I don't know how the model will perform though. 

Thanks for your support! Really appreciate it :)

**Reply by eikaramba** [reply]
i have the same problem. however i think i know the culprit at least. i have a labels.json with e.g. "ä". It is utf-8 encoded. i wrote a small script to output any characters which are not in my vocab set:
```py
import json

character_set = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~°£€¥¢฿äöüßÄÖÜẞ"

with open('./output/labels.json', 'r', encoding='utf-8') as file:
    data = json.load(file)

missing_characters_with_keys = {}

for key, value in data.items():
    for char in value:
        if char not in character_set:
            if char not in missing_characters_with_keys:
                missing_characters_with_keys[char] = [key]
            else:
                missing_characters_with_keys[char].append(key)

# Print the missing characters along with their corresponding keys
for char, keys in missing_characters_with_keys.items():
    print(f"Character '{char}' is missing from the character set and was found in the following keys: {', '.join(keys)}")
```

first i did it without the utf-8 encoding and saw the exact characters which doctr now complains about (after adding the print statements to the utils file as mentioned above. btw. please add this very valuable output by default). after reading the file with utf-8 encoding none characters were missing so my conclusion is that the labels.json file is just not read in the correct encoding format.

*update*: **Yes it works now** after editing `doctr\datasets\recognition.py` line 39:
```diff
        self.data: List[Tuple[str, str]] = []
-        with open(labels_path) as f:
+        with open(labels_path, encoding='utf-8') as f:
```

**Reply by felixdittrich92** [reply]
@eikaramba Would you like to open a PR to add the encoding ? :)

**Reply by eikaramba** [reply]
@felixdittrich92 will do :)

**Reply by eikaramba** [reply]
https://github.com/mindee/doctr/pull/1479

**Reply by durajetz** [reply]
Hey @tilman67,

since you shared your model with german vocab, would you mind telling me how can I use this model for my mindee inference script?

**Reply by felixdittrich92** [reply]
tilman-rassy/doctr-crnn-vgg16-bn-fascan-v1

```
from doctr.io import DocumentFile
from doctr.models import ocr_predictor, from_hub

img = DocumentFile.from_images(['<image_path>'])
# Load your model from the hub
model = from_hub('tilman-rassy/doctr-crnn-vgg16-bn-fascan-v1')

# Pass it to the predictor
# If your model is a recognition model:
predictor = ocr_predictor(det_arch='db_resnet50',
                          reco_arch=model,
                          pretrained=True)


# Get your predictions
res = predictor(img)
```

:)

**Reply by durajetz** [reply]
@felixdittrich92 thank you very much, this was quite helpful :)

**Reply by tilman67** [reply]
Hi @durajetz ,

sorry for being late, i've been quite busy the last weeks :-(
An thank you very much, @felixdittrich92 , for answering the question!

Best regards,
Tilman

**Reply by durajetz** [reply]
Hi @tilman67,
No problem mate, @felixdittrich92 responded very fast and was helpful.

Since you here, I am trying to train this same model `crnn_vgg16_bn ` with turkish vocabulary, in a high level with some minimal steps can you tell me what approach should I follow, since tbh I find rlly hard sometimes this mindee documentation :(

**Reply by tilman67** [reply]
Hi @durajetz ,

here is a short summary:

First, I created synthetic training data with the aid of [synthtiger](https://github.com/clovaai/synthtiger). This tool creates images from texts. To get the texts, I used free dictionaries from the web as well as automatically created character sequences. The latter contained a lot of german umlauts and other things I wanted the model to learn.

Next, I converted the synthtiger output to the format of the training data of doctr. This was essentially the format of PyTorch, since I used PyTorch.

Finally, I started the training, using the ``references/recognition/train_pytorch.py`` script with minor adaptions. The only difference was that the script had to know my vocabulary. I simply defined the vocab in the script and let the script use it if the ``--vocab`` option had a certain value. Furthermore, I set the ``--pretrained`` flag and reco_arch was ``crnn_vgg16_bn``.

Feel free to ask further questions!

Best regards,
Tilman

**Reply by durajetz** [reply]
@tilman67,

Thank you for your well explained summary.

I just have some questions regarding `synthtiger`.
I looked into the documentation of it but was not helpful at all, what format should corpuse be, right now I created a .txt file with 400k turkish words. Would you mind to provide me the synthtiger script to generate the images? and how to convert the output to PyTorch format?

Other than that, I don't want to disturb you any more, so thx alot

**Reply by tilman67** [reply]
Yes, the synthtiger documentation is not very detailed :-) It took me some time to get it working.

There is an ``examples/synthtiger`` directory in the synthtiger repo. I used that as a starting point. It's briefly described in the [docs](https://github.com/clovaai/synthtiger?tab=readme-ov-file#usage). My config file is very similar to ``config_horizontal.yaml``. I changed the path to the texts, the augmentation_charset (to meet my vocab with umlauts), and the so-called colormap, with I created before as described in the docs. Concerning the paths, I changed
```
corpus:
  weights: [1, 0]
  args:
    # length augmentable corpus
    - paths: [resources/corpus/mjsynth.txt, resources/corpus/synthtext.txt]
      weights: [1, 1]
```
to
```
corpus:
  weights: [1, 0]
  args:
    # length augmentable corpus
    - paths: [resources/corpus/fascan.txt]
      weights: [1]
```
``resources/corpus/fascan.txt`` is th file containing the texts from which images should be created.

  My ``template.py`` is the same as in the example. I only changed the Python class name.

The synthtiger call then looked like:
```
    synthtiger \
      -o "results" \
      -w 8 \
      -c "100000 \
      -v \
      "config/template.py" \
      "FaScan" \
      "config/config.yaml"
```
"FaScan" is the name of the template class mentioned above.

Synthtiger writes a file ``gt.txt`` in the output directory (``results`` in my case). If consists of lines of the form
```
<img_path> <text>
```
where ``<text>`` is the text and ``<img_path>`` is the path to the image created from the text. Thus, ``gt.txt`` provides a list of the synthtiger output.

Converting to doctr/PyTorch-Format can then be done as follows:

1. Choose a directory for the data.
2. Therein, create two subdirectories ``train``and ``val``.
3. In both ``train``and ``val``, create a subdirectory ``images``.
4. Parse the ``gt.txt`` file. Copy, say, 80%  of the images to ``train/images``. Create a JSON file ``labels.json`` in ``train``of the form
```
{
    "<img_filename_1>": "<text_1>",
    "<img_filename_2>": "<text_2>",
   ...
}
```
where <img_filename_1>, ... are the filenames of the images in ``train`` (without path) and <text_1>, ... the corresponding texts.
5. Do the same with the remaining 20% of the images, but with ``train`` replaced by ``val``.

That's it. Now the data is in the form required by doctr/PyTorch.

Hope that helps a bit....

If not, ask again :-)

Best regards,
Tilman

**Reply by durajetz** [reply]
Hey @tilman67, 
Firstly I want to thank you for this amazing explanation since it quite help me and I was able to generate the synth data successfully. There is another issue that I am encoutering now, and wanted to ask @felixdittrich92 maybe he knows more about this.

In **config_horizontal.yaml** file, I used the [tur.wordlist](https://github.com/tesseract-ocr/langdata_lstm/blob/main/tur/tur.wordlist)

> @tilman67, I have a question here, did you put a lot of word list for german in `corpus`, can you provide me any corpus example you used?

` - paths: [tur.txt]`
and
`augmentation_charset: tur_special.txt `
with this content:
```
(!"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~çğışöüÇĞİŞÖÜ)
```

and as font I used the same font @tilman67 did use and put it in the `synthtiger/resources/font/liberation` 

```
font:
  paths: [resources/font/liberation]
```

I generated 100k of synth data, and then converted into `train` (80%) and `val` (80%) for **Mindee/PyTorch** format.

Before starting the training, since I have old versions of nvidia drivers and wanted to use the training in GPU mode, I downgraded the PyTorch to **Version:** `1.12.1+cu113`, so this made me also update `train_pytorch.py` manually because this torch version does not support transformers.v2.

![image](https://github.com/mindee/doctr/assets/57491764/56f71383-9215-491d-84a7-e506107983fa)
![image](https://github.com/mindee/doctr/assets/57491764/3c10c838-d7ac-4c29-82b4-7450cc4e2fab)

Also in the `vocabs.py`, I added the turkish vocab like this:
`VOCABS["turkish"] = (VOCABS["english"] + "âçğışöüÂÇĞİŞÖÜ”©" + VOCABS["digits"] + VOCABS["punctuation"])`

After these adjustments, I started the training script with the following command:
```
python references/recognition/train_pytorch.py crnn_vgg16_bn --train_path train/ --val_path val/ --vocab "turkish" --pretrained --epochs 100 --name crn_vgg16_bn_turkish_v1 --workers 5
```
> My questions is here, should I also include in the training script the same fonts I generated the synth data?

And these are the loss results:
![image](https://github.com/mindee/doctr/assets/57491764/7c35865d-540d-41d9-accc-de62ff23e5f2)

But now when I test my model from this script:
```
import torch
from doctr.io import DocumentFile
from doctr.models import ocr_predictor, crnn_vgg16_bn
from doctr.datasets import VOCABS

reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False, vocab=VOCABS["turkish"])
reco_params = torch.load('./doctr/crn_vgg16_bn_turkish_v1', map_location="cpu")
reco_model.load_state_dict(reco_params)

print(reco_model.cfg['vocab'])

predictor = ocr_predictor(det_arch='db_resnet50',
                          reco_arch=reco_model,
                          pretrained=True)

doc = DocumentFile.from_images('turkish.jpg')

result = predictor(doc)

text = result.render()

print(text)
```
> The recognition is not that bad but it does not recognize the digits at all, why? 
Here is a small example:
![image](https://github.com/mindee/doctr/assets/57491764/22eed08f-2a59-4281-a3cb-41385f7545b9)
![image](https://github.com/mindee/doctr/assets/57491764/1ede7aa4-4ee6-4098-a0ea-d9cb1e340534)

**Reply by felixdittrich92** [reply]
Hi @durajetz :wave:,

1. Your vocab is wrong change to `VOCABS["turkish"] = VOCABS["english"] + "âçğışöüÂÇĞİŞÖÜ”©"` english contains already digits and punctuation
2. Please verify that your validation and training labels contains digits (The synthtiger charaugmenter was a bit broken in my case and worked not as expected as is) - best would be to create a char count over your labels to see also the distribution for each character inside your dataset

**Reply by felixdittrich92** [reply]
i have pushed my modified version to:
https://github.com/felixdittrich92/synthtiger/tree/doctr-modified

Be aware this is super dirty at the moment :sweat_smile:

**Comment by felixdittrich92** [comment]
Maybe it makes sense to share a modified fork from synthtiger which creates the dataset directly with the required doctr format @tilman67 wdyt ?
I have also used synthtiger to create the dataset for https://huggingface.co/Felix92/doctr-torch-parseq-multilingual-v1

**Reply by tilman67** [reply]
Yes, that may be helpful. If not a fork, at least a script to convert the data. Perhaps I can provide one. But I'll also think about the fork.

**Reply by felixdittrich92** [reply]
If there is a clean repo/modified fork we could link it in the documentation :) 
Unfortunately i don't have the time to work on that atm but if it helps i could also share my modifications

**Reply by durajetz** [reply]
Hey @felixdittrich92 

I saw that inside `synthtiger/resources/font` you have put different fonts. My question is, when you trained these synth data for parseq multilingual model, did you specify all of these fonts also in the training command?

**Reply by eikaramba** [reply]
these are just different ways to generate synth data. either by using synthtiger or by using doctrs wordgenerator (which is invoked via the training command). if you use the commands in the training step it will not use external files, you can also just look at the training script yourself and see how it works
